keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -0,0 +1,379 @@
+
+#### Story 2
+User will will no longer have to inject the Pull Always Image Pull Policy to
+ensure all tenants have rights to the images that are already present on a host.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+With the default of the feature gate being off, users / cloud providers will have
+to set the feature gate to true to gain these this Secure by Default benefit.
+
+### Risks and Mitigations
+
+Image authentications with a registry may expire. To mitigate expirations a
+a timeout could be used to force re-authentication. The timeout could be a
+container runtime feature or a `kubelet` feature. If at the container runtime,
+images would not be present during the EnsureImagesExist step, thus would have
+to be pulled and authenticated if necessary.
+
+Since images can be pre-loaded, loaded outside the `kubelet` process, and
+garbage collected.. the list of images that required authentication in `kubelet`
+will not be a source of truth for how all images were pulled that are in the
+container runtime cache. To mitigate, images can be garbage collected at boot.
+
+
+## Design Details
+
+Kubelet will track, in memory, a hash map for the credentials that were successfully used to pull an image. The hash map
+will not be persisted to disk, in alpha. For alpha explicitly, we will not reuse or add other state manager concepts to kubelet.
+
+See PR for detailed design / behavior documentation.
+
+### Test Plan
+
+For alpha, exhaustive Kubelet unit tests will be provided. Functions affected by the feature gate will be run with the feature gate on and with the feature gate off. Unit buckets will be provided for:
+- HashAuth - (new, small) returns a hash code for a CRI pull image auth [link](https://github.com/kubernetes/kubernetes/pull/94899/files#diff-ca08601dfd2fdf846f066d0338dc332beddd5602ab3a71b8fac95b419842da63R704-R751)
+- shouldPullImage - (modified, large sized change) determines if image should be pulled based on presence, and image pull policy, and now with the feature gate on if the image has been pulled/ensured by a secret. A unit test bucket did not exist for this function. The unit bucket will cover a matrix for:
+```
+	pullIfNotPresent := &v1.Container{
+    ..
+   	ImagePullPolicy: v1.PullIfNotPresent,
+ 	}
+ 	pullNever := &v1.Container{
+    ..
+    ImagePullPolicy: v1.PullNever,
+ 	}
+ 	pullAlways := &v1.Container{
+ 		..
+    ImagePullPolicy: v1.PullAlways,
+ 	}
+ 	tests := []struct {
+ 		description       string
+ 		container         *v1.Container
+ 		imagePresent      bool
+ 		pulledBySecret    bool
+ 		ensuredBySecret   bool
+ 		expectedWithFGOff bool
+ 		expectedWithFGOn  bool
+ 	}
+```
+[TestShouldPullImage link](https://github.com/kubernetes/kubernetes/pull/94899/files#diff-7297f08c72da9bf6479e80c03b45e24ea92ccb11c0031549e51b51f88a91f813R311-R438)
+
+At beta we should revisit if integration buckets are warranted for e2e node and/or cri-tools/critest, and after gathering feedback.
+
+### Graduation Criteria
```

Comment:
explicit beta criteria are not yet specified, but i would desire this function to work across kubelet restarts and node reboots, which would require some cached state persisted to disk.